### ISSUE TITLE
Update the connect message view chat after receiving a message

### ISSUE THREAD

**Issue by riverbaymark** [carekit-apple/CareKit#219]
Greetings,

So when you send a message there is a delegate function that is called (ex: func connectViewController(_ viewController: OCKConnectViewController, didSendConnectMessage message: String, careTeamContact contact: OCKContact)) where it is pretty easy to stick code to update an array of message items and the chat table is automatically updated.   The issue I am having is how do you trigger an update of this chat table when a message is received.  I am able to detect through a closure when a message is received, but how do I update the chat table if it is currently open?  This isn't an issue as long as someone isn't already in the chat view as it displays all the messages when the inbox is opened.  

Thanks!

Mark

**Comment by kelseydedoshka-apple** [comment]
@riverbaymark - great question. A couple initial suggestions that may help you. You could listen to notifications (maybe add a listener in viewDidLoad) in the OCKConnectMessagesViewController and whenever you get new messages post a notification to trigger that listener and in the listener callback trigger a table refresh.

In general the tableView displaying the messages just needs to be reloaded. So you could also expose a method in the header that allows you to refresh the table whenever necessary.

Great catch - definitely something we should build into the framework on our side as well. Hopefully this helps, please reach back out if this does not help resolve your issues.

**Comment by riverbaymark** [comment]
Thanks @kelseydedoshka-apple.  I will give that a try.

**Comment by kelseydedoshka-apple** [comment]
@riverbaymark - great, would love to know how that works out for you!

**Comment by kelseydedoshka-apple** [comment]
@riverbaymark - just wanted to check back in here and see if this helped solve your issue?

**Comment by riverbaymark** [comment]
Hi @kelseydedoshka-apple sorry have been bogged down writing API calls for some EHRs.  I am going to try this out this week as we just met some clients last pm requesting such functionality.  

On a side note and something to think about, as a provider myself and in talking with several other providers recently many are reluctant to have open messaging.  You have the greatest majority of patients that will never abuse such a feature, but a handful see this as an open hotline and will bombard you 24/7 which is not appropriate and can bring in some legal responsibility issues for the provider.  Something that is highly recommended is a way to use charts / graphs in combination with notifications that a provider can see when thresholds are crossed.  Sort of like a red zone (danger), yellow zone (warning), green zone (all's good) type scenario.  I know that description is a little abstract but would be glad to provide some concrete examples.  Just something to think about maybe for a "ProviderKit".

Thanks and I will try to get some information back to you very soon!

Mark

**Comment by kelseydedoshka-apple** [comment]
@riverbaymark - not a problem, thanks for getting back to me here! This is a great suggestion and I think others share similar ideas as you do here with a way for providers to see information - thanks for sharing your thoughts on this. 

As this is a much larger concept I would be curious to also chat about maybe a smaller solution to the challenge you mention about patients abusing the open messaging. I am wondering if there is a way we could provide some sort of limitations onto the messaging module - perhaps like access levels or something? To where patients could only message a certain number of times, or if someone initiates the conversation with them? Would love to hear your thoughts on something like this!

**Comment by riverbaymark** [comment]
@kelseydedoshka-apple Sorry for the delay!  This brings up interesting questions.  
There are concierge medicine practices and other practices that offer enhanced services that promote direct access to a provider, so some type of access control would be a great idea.  In this scenario it is likely a patient could initiate a conversation but the provider should have the ability to "close" it.  
The ability for a provider to initiate a conversation with a patient and then "close" it is also an excellent idea.
Traditionally offices have a "phone nurse" or "message center" where everything is routed through.  Questions are then brought to the provider and an answer is relayed back to the patient.  I could almost envision this as a "group message" where a message stays open until it is answered by an appropriate party with multiple persons (nurses, providers, etc.) being able to view it.  
The hard thing is with a continuous message stream, if a patient has multiple interactions, then being able to go back and see specific items is hard over time.  There almost needs to be a way to have the message interaction but it be categorized by person (primary) and by issue (secondary).  One great analogy I can think of is this very format we are discussing things in right now.  In GitHub there are issues a person brings up, someone can answer the issue at hand, and then it can be closed but looked back on historically if need be (as often is the case).  In the GitHub example though it is kinda the reverse in that issue is primary and the person asking the question is secondary.  

To summarize all of this I think patients being able to initiate contact with a clinic that is open to receive the messages is a good thing (like the group message type idea above).  Provider initiated direct contact with a patient with ability to close would be a good thing except in the concierge type cases mentioned above the provider allows the patient to initiate contact.  

Sorry for the long winded answer, but hope it makes some sense.

Thanks!

Mark

**Comment by kelseydedoshka-apple** [comment]
@riverbaymark - this is all great information I really appreciate you sharing it all! I think we are both in agreement that some kind of access control could be really useful here. Would you be open to creating a new issue with this request (access controls on messages), and we can mark it as a new feature request so others can chime in on the conversation or consider developing it out?

**Comment by riverbaymark** [comment]
@kelseydedoshka-apple Absolutely!  Would be happy to do that.

**Comment by kelseydedoshka-apple** [comment]
@riverbaymark - great thanks! Ok to close this issue out and continue the discussion on the other thread?

**Comment by riverbaymark** [comment]
@kelseydedoshka-apple Certainly!
